pageserver/src/walreceiver.rs

```diff
@@ -27,56 +27,55 @@ use crate::config::PageServerConf;
 use crate::task_mgr::WALRECEIVER_RUNTIME;
 
 use anyhow::{ensure, Context};
-use etcd_broker::Client;
-use itertools::Itertools;
+use neon_broker::neon_broker_proto::neon_broker_client::NeonBrokerClient;
 use once_cell::sync::OnceCell;
 use std::future::Future;
 use tokio::sync::watch;
+use tonic::transport::Channel;
 use tracing::*;
-use url::Url;
 
 pub use connection_manager::spawn_connection_manager_task;
 
-static ETCD_CLIENT: OnceCell<Client> = OnceCell::new();
+static BROKER_CLIENT: OnceCell<NeonBrokerClient<Channel>> = OnceCell::new();
 
 ///
-/// Initialize the etcd client. This must be called once at page server startup.
+/// Initialize the broker client. This must be called once at page server startup.
 ///
-pub async fn init_etcd_client(conf: &'static PageServerConf) -> anyhow::Result<()> {
-    let etcd_endpoints = conf.broker_endpoints.clone();
+pub async fn init_broker_client(conf: &'static PageServerConf) -> anyhow::Result<()> {
+    let broker_endpoints = conf.broker_endpoints.clone();
     ensure!(
-        !etcd_endpoints.is_empty(),
-        "Cannot start wal receiver: etcd endpoints are empty"
+        !broker_endpoints.is_empty(),
+        "Cannot start wal receiver: broker endpoints are empty"
     );
 
-    let etcd_client = Client::connect(etcd_endpoints.clone(), None)
+    // Note: we do not attempt connecting here (but validate endpoints sanity).
+    let broker_client = NeonBrokerClient::connect_lazy(broker_endpoints.clone())
         .await
-        .context("Failed to connect to etcd")?;
-
-    // FIXME: Should we still allow the pageserver to start, if etcd
-    // doesn't work? It could still serve GetPage requests, with the
-    // data it has locally and from what it can download from remote
-    // storage
-    if ETCD_CLIENT.set(etcd_client).is_err() {
-        panic!("etcd already initialized");
+        .context(format!(
+            "Failed to connect to broker at {}",
```

Comment:
That's not true anymore, since we connect lazily?

Also, is it really beneficial to connect lazily? This way we at least fail fast and bring down the entire pageserver with us, while with the new scenario, we could start and pretend to be working and then start to fail in walreceiver, which is harder to test and understand (I would imagine, the client would see some WAL timeout error, as usual, if this happens?)

Reply:
We actually always connected lazily. Initial client creation only checked that endpoints are well formed -- and this is still the case, I believe. And I do think that's the right way -- requirement for broker to be available when pageserver starts is obscure and not needed. For example, currently you won't be even able to run tests, as we start pageserver first and broker second.